### PR TITLE
userland: allow/enforce compile-time STACK_SIZE

### PR DESCRIPTION
**Pull request by ppannuto**
### Pull Request Overview

Change the use of STACK_SIZE in crt1 to be an external variable.
We were already defining this as a symbol during linking, but we'd
also defined it at compile-time for libtock, so it was fixed.
This fixes to actually resolve at link time and further to protect
against libtock accidentally compiling with a fixed stack size in
the future.

@adkinsjd this should fix part of the issue you were having. It'll
allow your applications to change the stack size, but it will not
automatically rebuild if you only change stack size, you'll need
to touch any other source file so that the app relinks.

### Testing Strategy

Compile-tested only.

### TODO or Help Wanted

Testing, or I'll do it tomorrow.

~~Also TODO: It occurs to me as I make this PR that I should look into the userland documentation to make sure it explains STACK_SIZE well.~~

### Documentation Updated

- [x] Kernel: Updated the relevant files in `/docs`, or no updates are required.
- [x] Userland: Added/updated the application README, if needed.

### Formatting

- [x] Ran `make formatall`.

### CODE REVIEW

**Comment by ppannuto** [comment]
[Meta]: Do we think priorities / review procedure should apply to userland-only changes as well? This touches literally every app (though it should be in a pretty benign way) so I've marked it `P-Significant` for now.

**Comment by phil-levis** [comment]
This does seem like a significant change -- modifications to process layouts/code shared across all processes should probably be considered Significant by default.

**Comment by ppannuto** [comment]
STACK_SIZE needs to be defined at _link_ time and needs to _not_ be defined at compile time.

Previously, we were defining it during both compilation and linking (the latter was thus ignored). This PR removes the compile-time definition. It leaves the [already existing link-time definition](https://github.com/helena-project/tock/blob/master/userland/AppMakefile.mk#L141) alone.

**Comment by ppannuto** [comment]
Okay, this should be good to go now.

I actually had to add the stack size to the header that the linker prepares. The original design of using a symbol defined by the linker meant that it was a r9-relative load, but this early in initialization, we haven't setup relocations yet (indeed, the initial value of r9 is based in part off the stack size), so it needs to be at a fixed location instead.

Tested fairly extensively on hail with lots of different apps and the various mpu test apps.

File: doc/Userland.md [hunk 1]
```diff
@@ -272,6 +272,13 @@ Applications can specify their required stack and heap sizes by defining the
 make variables `STACK_SIZE` and `APP_HEAP_SIZE`, which default to 2K and 1K
 respectively as of this writing.
 
+`libtock` will set the stack pointer during startup. To allow each application
+to set its own stack size, `libtock` expects an external symbol `STACK_SIZE` to
```
Comment (bradjc) [inline doc/Userland.md]: This is no longer true, correct? The value is in the crt0 header, and the userland app shouldn't expect to find `STACK_SIZE` defined.

Comment (ppannuto) [reply doc/Userland.md]: This is wrong, I'll update it now. The value is not in the crt0 header (there isn't one anyone, that was a failed experiment), it's defined during build and placed in the application header by the linker script.

**Comment by ppannuto** [comment]
@helena-project/core-team let's start the 7 day clock

@phil-levis is your question resolved?

**Comment by bradjc** [comment]
@phil-levis Bump, because I'd like to incorporate this into the elf2tbf refactor.

**Pull request merged**